### PR TITLE
Add cascading config finder

### PR DESCRIPTION
**Pull request by isbm**
Boxed plugin can only get a static hard-coded configuration, delivered as a resource inside a Python package. It cannot be overridden without explicit change of a package-delivered data. This is also makes it impossible to programmatically integrate it with Kiwi as a library.

Under these circumstances, a cascading mechanism is required, which is searching for the configuration file in the following order:
1. Environment variable "KIWI_BOXED_PLUGIN_CFG" (full path, the name can be anything)
2. Current directory, from where a software was called
3. `$HOME/.config/kiwi/kiwi_boxed_plugin.yml`
4. `/etc/kiwi_boxed_plugin.yml`
5. Resource (default config, coming with the package)

This PR shares https://github.com/Elektrobit/kiwi-boxed-plugin/pull/1 and is remake of https://github.com/OSInside/kiwi-boxed-plugin/pull/58

### CODE REVIEW

**Comment by isbm** [comment]
@schaefi had to re-PR...

File: .gitignore [hunk 1]
```diff
@@ -77,3 +77,6 @@ doc/xml/
 
 # man pages
 *.1.gz
+
+# VS
+.vscode/
```
Comment (schaefi) [inline .gitignore:82]: empty vscode settings ?

**Comment by schaefi** [comment]
And the expected coverage issue for the new code paths not yet covered by a test
```
/home/runner/work/kiwi-boxed-plugin/kiwi-boxed-plugin/kiwi_boxed_plugin/defaults.py                   65      4    94%   55, 60, 65, 69
```
If you can provide a unit test for the new search order all test will go green and we can merge

Thanks

File: .gitignore (hunk 1 above)
Comment (isbm) [reply .gitignore:82]: It simply ignores the entire `.vscode/` dir.

Comment (schaefi) [reply .gitignore:82]: Ah ok, you are coding in vscode have your own settings and don't want to be bothered by their contents, got it. Unrelated to the proposed changed but doesn't hurt, ok

File: kiwi_boxed_plugin/defaults.py [hunk 2]
```diff
@@ -37,8 +38,38 @@ class Defaults:
...
+        cfg_p: str | None = os.environ.get("KIWI_BOXED_PLUGIN_CFG")
+        if cfg_p is not None and os.path.exists(cfg_p):
+            return cfg_p
+
+        # Curr dir
+        cfg_p = os.path.abspath(cfg_n)
+        if os.path.exists(cfg_p):
+            return cfg_p
+
+        # Local Kiwi config
```
Comment (schaefi) [inline kiwi_boxed_plugin/defaults.py:62]: Can we change this comment to ```User specific Kiwi config``` ?

File: kiwi_boxed_plugin/defaults.py [hunk 3]
```diff
@@ -37,8 +38,38 @@ class Defaults:
...
+        cfg_pth: pathlib.Path = pathlib.Path.home().joinpath(f".config/kiwi/{cfg_n}")
+        if cfg_pth.exists():
+            return cfg_pth.as_posix()
+
+        cfg_p = f"/etc/{cfg_n}"
```
Comment (schaefi) [inline kiwi_boxed_plugin/defaults.py:67]: For consistency, can we add a comment e.g ```# system wide Kiwi config``` ?

File: kiwi_boxed_plugin/defaults.py [hunk 4]
```diff
@@ -37,8 +38,38 @@ class Defaults:
...
+        if os.path.exists(cfg_p):
+            return cfg_p
+
         return resource_filename(
```
Comment (schaefi) [inline kiwi_boxed_plugin/defaults.py:71]: For consistency can we add a comment e.g ```# module provided Kiwi config``` ?

**Review by schaefi — COMMENTED** [review]
Just nitpick. I think the proposed comment changes helps to identify the code block for the different search strategies faster. Alternatively it can be done by more descriptive variable names. I guess ```cfg_p``` stands for ```config path``` and ```cfg_n``` for ```config name```. I leave this up to you but something like

```python
config_current_path: str = os.path.abspath(config_name)
if os.path.exists(config_current_path):
    return config_current_path

config_user_path: pathlib.Path = pathlib.Path.home().joinpath(f".config/kiwi/{cfg_name}")
if config_user_path.exists():
    return config_user_path.as_posix()

... and so on
```
prevents you from the extra comments and makes the code more explicit in my eyes.

Just an idea

**Comment by isbm** [comment]
Alternatively, you can refactor it yourself the way you like.

**Comment by schaefi** [comment]
> Alternatively, you can refactor it yourself the way you like and I would rather prefer that.

true, in this case the only missing part are the tests for the code you added. Thanks

**Comment by isbm** [comment]
Fair enough, I will add them ASAP.

**Pull request closed without merging**